runtime/compiler/optimizer/UnsafeFastPath.cpp

```diff
@@ -49,6 +49,126 @@
+   if (!performTransformation(comp(), "%s turning the call [" POINTER_PRINTF_FORMAT "] into atomic intrinsic\n", optDetailString(), node))
+      return;
+
+    TR::MethodSymbol *symbol = node->getSymbol()->castToMethodSymbol();
+   // Codegen will inline the call with the flag
+   //
+   if (symbol->getMethod()->isUnsafeCAS(comp()))
+      {
+      // codegen doesn't optimize CAS on a static field
+      //
+      if (callerMethod != TR::java_lang_invoke_StaticFieldVarHandle_StaticFieldVarHandleOperations_OpMethod)
+         {
+         node->setIsSafeForCGToFastPathUnsafeCall(true);
+         if (trace())
+            traceMsg(comp(), "Found Unsafe CAS on non-static field, set the flag\n");
+         }
+
+      return;
+      }
+
+   TR::SymbolReferenceTable::CommonNonhelperSymbol helper = TR_J9MethodBase::equivalentAtomicIntrinsic(calleeMethod);
+   if (!comp()->cg()->supportsNonHelper(helper))
+      {
+      if (trace())
+         traceMsg(comp(), "Equivalent atomic intrinsic is not supported on current platform, quit\n");
+      return;
+      }
+
+   // Give up on arraylet
+   //
+   if (TR_J9MethodBase::isVarHandleOperationMethodOnArray(callerMethod)
+       && TR::Compiler->om.canGenerateArraylets())
+      return;
+
+   TR::Node* unsafeAddress = NULL;
+   if (callerMethod == TR::java_lang_invoke_StaticFieldVarHandle_StaticFieldVarHandleOperations_OpMethod)
+      {
+      TR::Node *jlClass = node->getChild(1);
+      TR::Node *j9Class = TR::Node::createWithSymRef(node, TR::aloadi, 1, jlClass, comp()->getSymRefTab()->findOrCreateClassFromJavaLangClassSymbolRef());
+      TR::Node *ramStatics = TR::Node::createWithSymRef(node, TR::aloadi, 1, j9Class, comp()->getSymRefTab()->findOrCreateRamStaticsFromClassSymbolRef());
+      TR::Node *offset = node->getChild(2);
+      // The offset for a static field is low taged, mask out the flag bit to get the real offset
+      //
+      offset = TR::Node::create(node, TR::land, 2, offset,
+                                                   TR::Node::lconst(node, ~J9_SUN_FIELD_OFFSET_MASK));
+      unsafeAddress = TR::Compiler->target.is32Bit() ? TR::Node::create(node, TR::aiadd, 2, ramStatics, TR::Node::create(node, TR::l2i, 1, offset)) :
+                                                       TR::Node::create(node, TR::aladd, 2, ramStatics, offset);
+      }
+   else
+      {
+      TR::Node* object = node->getChild(1);
+      TR::Node* offset = node->getChild(2);
+      unsafeAddress = TR::Compiler->target.is32Bit() ? TR::Node::create(node, TR::aiadd, 2, object, TR::Node::create(node, TR::l2i, 1, offset)) :
+                                                       TR::Node::create(node, TR::aladd, 2, object, offset);
+      unsafeAddress->setIsInternalPointer(true);
+      }
+
+
+   if (callTree->getNode()->getOpCode().isNullCheck())
+      {
+      TR::Node::recreate(callTree->getNode(), TR::treetop);
+      if (trace())
+         traceMsg(comp(), "Remove null check on the call\n");
```

Comment:
node number and explanation why

Reply:
Modified the code to preserve the check. VP will fold the check away since final fields in java/lang/invoke is always foldable.